### ISSUE TITLE
Ability to extend post with extra fields and extra placeholders

### ISSUE THREAD

**Issue by g3rb3n** [nephila/djangocms-blog#569]
We would like to add some properties to a blog post which should be manageable in the admin interface and in the web UI.

We currently have a monkeypatch to achieve this. The extra properties are added as an inline to the PostAdminForm, extra placeholders are auto created with signals.

It would be nice if there are some hooks to achieve this. eg:
def register_post_inline(inline)
def register_post_placeholder(model)

This enable any one to create a model and admin like:
class PostExtension(models.Model):
    post = models.ForeignKey(Post, on_delete=models.CASCADE, related_name='extension')
    some_field = models.CharField(max_length=10)

class PostExtensionInline(admin.TabularInline):
    model = PostExtension
    fields = ['some_field']
    classes = ['collapse']
    extra = 1
    can_delete = False

def register_post_inline(inline)

{{ post.extension.some_field }}

And:
class PostPlaceholderExtension(models.Model):
    post = models.OneToOneField(Post, on_delete=models.CASCADE, related_name='placeholders')
    some_placeholder = PlaceholderField('some_placeholder')

register_post_placeholder(PostPlaceholderExtension)

{% render_placeholder post.placeholders.some_placeholder %}

@yakky What do you think? Are inlines a good solution to achieve this? Or do you have better ideas?

**Comment by yakky** [comment]
I am not against it, but I am a bit worried about the number of features this application is growing :joy: 
Can't this already be achiedeved by creating external models pointing to the post and overriding the admin? :thinking:

**Comment by g3rb3n** [comment]
@yakky The idea is not to include to many features, but giving programmers a generic ability to hook up features that should not be part of a generic blog application.

Some want to include images, some not. 
Some want to include extra placeholder, some not.

We have someone that wants some fields which are very specific and should not be in a generic blog application. We already achieved the functionality with a monkeypatch and a signal. I also looked at overriding the admin form, but encountered some issues with the fieldsets. So the monkey patch with a inline seems the most straightforward way to achieve this. Thought it would be nice to offer this functionality in a more generic way.

I will write a proposal. If you think it is to much to maintain we will stick with the current solution.

**Comment by yakky** [comment]
@g3rb3n I got the idea to just implement hooks to make adding things easier, I was just wondering if it's achievable in a clean way with the current codebase or not, but I have not analyzed in great detail, yet

If you can write a detailed (even if brief) proposal, I think it would be an interesting addition

**Comment by g3rb3n** [comment]
I have the required changes here:
https://github.com/g3rb3n/djangocms-blog/tree/feature/issue-569-post-extensions

Basically include `inlines = []` in admin.PostAdmin

And the following:
```
def register_post_inline_extension(inline):
    """ Register a admin.Inline to the Post form.
        Model should have a post = models.OneToOneField(Post)
    """
    PostAdmin.inlines = type(PostAdmin.inlines)([inline]) + PostAdmin.inlines


def unregister_post_inline_extension(inline):
    """ Unregister a previously regsitered admin.Inline to the Post form.
    """
    PostAdmin.inlines.remove(inline)


def register_post_placeholder_extension(model):
    """ Register a model with placeholders to be auto created when a post is created.
        Model should have a post = models.OneToOneField(Post)
    """
    signals.post_save.connect(
        create_post_post_save(model), sender=Post, weak=False,
    )


def create_post_post_save(model):
    """ A wrapper for creating create_instance function for a specific model.
    """

    def create_instance(sender, instance, created, **kwargs):
        """ Create Model instance for every new Post.
        """
        if created:
            model.objects.create(post=instance)

    return create_instance
```

**Comment by yakky** [comment]
@g3rb3n thanks, now it's clearer :)

Generally speaking I think it's :+1: 

I am thinking if we need two different code paths (inlines and onetoone), or we could just go with one and cover all use cases with that, just to make things easier to understand for the people who would need this feature.

**Comment by g3rb3n** [comment]
I agree that a register hook should be easy to understand. There is a functional difference between the inline and the placeholder. The inline should appear in the form and only has to be created if specified. The placeholder is not in the admin and should always be created if a post is created.

It might be best to communicate this difference clearly in the docs. But maybe there is a clean way to cover both. Signals together with inlines do not work well and introduce edge cases. That should be avoided.

I will look into this!

**Comment by yakky** [comment]
@g3rb3n let me sit on this and think a possible unified solution
If I don't come up with anything, your implementation is :ok_hand:

**Comment by yakky** [comment]
I think we have no way around the two different models, but I suggest to slightly refactor the code.

The current `placeholder_extension` is not required to have a placeholder and it might register an `inline` itself, to allow adding normal fields and not exclusively placeholder fields, thus the functions might be adapted (pseudocode) this way:

```python
def register_post_extension(inline):
    """ Register a admin.Inline to the Post form.
        Model should have a post = models.ForeignKey(Post) or post = models.OneToOneField(Post)
    """
    PostAdmin.inlines = type(PostAdmin.inlines)([inline]) + PostAdmin.inlines
    .if inline.model. has onetoone:
        signals.post_save.connect(
            create_post_post_save(model), sender=Post, weak=False,
        )


def unregister_post_inline_extension(inline):
    """ Unregister a previously regsitered admin.Inline to the Post form.
    """
    PostAdmin.inlines.remove(inline)
    .if inline.model. has onetoone:
        signals.post_save.disconnect(
            create_post_post_save
        )
```

This provides the same function for both, and you only need to decide if you want a `OneToOne` or a `ForeignKey`

What do you think?

**Comment by g3rb3n** [comment]
The thing is though that inlines are handled by django and django automatically creates related instances when the admin user specifies the field. When a admin user leaves the field empty the placeholder is also not available nor createable.

The best would be to let django always create the related instance, also when u user does not specify the inline. Than we can also leave out the signals.

I will try this out soon.

**Comment by yakky** [comment]
@g3rb3n sorry, my pseudocode above probably was not  clear enough
fields with one-to-one must still be created via signal, but it can be handled in the same register function, by checking if the model linked to the inline has a `OneToOneField` to the post

By always adding the `InlineAdmin` whatever the key is, we also allow users to add fields on the model with the `OneToOneField`

**Comment by g3rb3n** [comment]
So basically one register function and in the function the check on whether this instance should be created with a signal or not. That idea is clear.

There are two situations:
- admin user specifies the admin inline -> object is created by django
- admin does not specify the inline -> object with placeholder should be created with a signal

placeholder can always be null and are filled in later in the cms ui.
standard django fields can be mandatory, objects should only be created if a user specifies the inline instance.

Or am I missing something here?

**Comment by yakky** [comment]
It's correct, do you think it's doable? I *think* that inlines for `OneToOneField` are always created on save, even if blank, but I have not tested it

**Comment by g3rb3n** [comment]
I have refactored the code so that there is just one register_extension function. I have also added unregister functionality.

Objects in inlines are not always created as far as I know. If left blank, this might be the expected behavior.

If someone wants to add just a placeholder an inline does not make sense, since placeholders are only visible in the CMS UI, not in the django admin UI. This is now supported, someone can create an OneToOne model and register this model as an extennsion.

I have created two PR's. One based upon the develop branch and one based upon the feature/issue-564-update-tooling branch.

If you have any comments, please let me know.

**Comment by yakky** [comment]
@g3rb3n awesome! I will review them quickly. #568 has already been merged on develop, BTW so we only need one

**Comment by yakky** [comment]
Fixed by #590
